lib/routes/account.js

```diff
@@ -1026,37 +1073,88 @@ module.exports = function (
         }
       },
       handler: function (request, reply) {
-        log.begin('Account.RecoveryEmailVerify', request)
         var uid = request.payload.uid
         var code = Buffer(request.payload.code, 'hex')
+        var tokenData = {uid: uid}
+
+        log.begin('Account.RecoveryEmailVerify', request)
         db.account(Buffer(uid, 'hex'))
           .then(
             function (account) {
-              // If the account is already verified, they may be e.g.
-              // clicking a stale link.  Silently succeed.
-              if (account.emailVerified) {
-                log.increment('account.already_verified')
-                return true
-              }
-              if (!butil.buffersAreEqual(code, account.emailCode)) {
-                throw error.invalidVerificationCode()
-              }
-              log.timing('account.verified', Date.now() - account.createdAt)
-              log.event('verified', { email: account.email, uid: account.uid, locale: account.locale })
-              log.increment('account.verified')
 
-              // send a push notification to all devices that the account changed
-              push.notifyUpdate(uid)
+              /**
+               * This describes the logic behind verify_code and how it treats different requests.
+               *
+               * 1) Handling unverified email account
+               *
+               *    Check to see if `code` is the same as the account.emailCode
+               *
+               *    A) If it is, then this is a request to verify an account email. Verify email and
+               *       tokens associated with sessions.
+               *
+               *    B) Else this is a request to verify tokens, with an unverified account. `code` is a `tokenVerificationId`
+               *       Attempt to verify tokens, if `tokenVerificationId` is not found, fail request. If `tokenVerificationId`
+               *       is found, verify sessions and verify account email.
+               *
+               * 2) Handling verified email account
+               *
+               *    The request is coming from an already verified user account, it only can be a request to
+               *    verify tokens. The `code` param is a tokenVerificationId.
+               */
+              if (!account.emailVerified) {
```

Comment:
For any unverified account email, a check is done to verifyTokens and email, so this can definitely be made a little more concise and simplified.